### PR TITLE
MULE-17741: FirstSuccessful implementation is blocking

### PR DESCRIPTION
**Pull request by mbuchwald**

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/internal/routing/FirstSuccessfulRouter.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+import static java.util.Optional.*;
+import static org.mule.runtime.api.functional.Either.left;
+import static org.mule.runtime.api.functional.Either.right;
+import static org.mule.runtime.core.internal.event.EventQuickCopy.quickCopy;
+import static org.mule.runtime.core.privileged.processor.MessageProcessors.applyWithChildContext;
+import static org.slf4j.LoggerFactory.getLogger;
+import static reactor.core.Exceptions.propagate;
+import static reactor.core.publisher.Mono.subscriberContext;
+
+public class FirstSuccessfulRouter {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/routing/FirstSuccessfulRouter.java]: package visibility

File: core/src/main/java/org/mule/runtime/core/internal/routing/FirstSuccessfulRouter.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+import org.mule.runtime.core.internal.event.EventInternalContextResolver;
+import org.mule.runtime.core.internal.exception.MessagingException;
+import org.mule.runtime.core.internal.message.InternalEvent;
+import org.mule.runtime.core.internal.rx.FluxSinkRecorder;
+import org.mule.runtime.core.privileged.routing.CouldNotRouteOutboundMessageException;
+import org.reactivestreams.Publisher;
+import org.slf4j.Logger;
+import reactor.core.publisher.Flux;
+
+import java.util.*;
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/routing/FirstSuccessfulRouter.java]: nope

File: core/src/main/java/org/mule/runtime/core/internal/routing/FirstSuccessfulRouter.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+  // this leads to a reactor completion signal being emitted while the event is being re-injected for retrials. This is solved by
+  // deferring the downstream publisher completion until all events have evacuated the scope.
+  private final AtomicInteger inflightEvents = new AtomicInteger(0);
+  private final AtomicBoolean completeDeferred = new AtomicBoolean(false);
+
+  public FirstSuccessfulRouter(Component owner, Publisher<CoreEvent> publisher, List<ProcessorRoute> routes) {
+    this.owner = owner;
+
+    innerRecorders =
+        IntStream.range(0, routes.size()).mapToObj(x -> new FluxSinkRecorder<CoreEvent>()).collect(Collectors.toList());
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/routing/FirstSuccessfulRouter.java]: why use an `IntStream` instead of just `routes.stream()`?

File: core/src/main/java/org/mule/runtime/core/internal/routing/ScatterGatherRouter.java [hunk 4]
```diff
@@ -95,4 +98,18 @@ protected ForkJoinStrategyFactory getDefaultForkJoinStrategyFactory() {
...
+
+  /**
+   * Validates that the payload is not consumable so it can be copied.
+   *
+   * If validation fails then throws a MessagingException
+   *
+   * @param message
+   * @throws MuleException if the payload is consumable
+   */
+  public static void validateMessageIsNotConsumable(Message message) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/routing/ScatterGatherRouter.java]: is this change related?

Comment (mbuchwald) [reply core/src/main/java/org/mule/runtime/core/internal/routing/ScatterGatherRouter.java]: ScatterGather was using that method from one of the removed classes (yeah, it depended on a class of first successful 🤦‍♂ )

File: core/src/main/java/org/mule/runtime/core/internal/routing/FirstSuccessfulRouter.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+  private void completeRouterIfNecessary() {
+    if (completeDeferred.get() && inflightEvents.get() == 0) {
+      completeRouter();
+    }
+  }
+
+  /**
+   * Assembles and returns the downstream {@link Publisher<CoreEvent>}.
+   *
+   * @return the successful {@link CoreEvent} or retries exhaustion errors {@link Publisher}
```
Comment (mbuchwald) [inline core/src/main/java/org/mule/runtime/core/internal/routing/FirstSuccessfulRouter.java]:
```suggestion
   * @return the successful {@link CoreEvent} or routing errors {@link Publisher}
```
